### PR TITLE
hw/mcu/nordic/nrf52xxx: Using hfxo_request and release. Addresses #2393.

### PR DESCRIPTION
**Pull request by ncasaril**
hal_timer and hal_system_clock_start both enabled the HFXO directly. 
Now they use the nrf52_clock_hfxo_request, and where suitable also nrf52_clock_hfxo_release.

### CODE REVIEW

File: hw/mcu/nordic/nrf52xxx/src/hal_timer.c
Comment (kasjer) [inline hw/mcu/nordic/nrf52xxx/src/hal_timer.c:666]: **#endif** was lost in the process

Comment (ncasaril) [reply hw/mcu/nordic/nrf52xxx/src/hal_timer.c:666]: Thanks @kasjer. 

In the hal_timer_deinit I've added the same ifdef around the release and also checking that it's not timer 5 to match hal_timer_config. 

In the deinit, I realise that the if-statement checks for tmr_rtc which is only enabled for timer_num=5 and so the release could go into the else section there. However using the timer_num as in hal_timer_config may be easier to understand?

**Comment by kasjer** [comment]
@ncasaril this patch slightly changes how code used to work.
Before it waited for XTAL to settle now it just starts it without waiting.
I don't know the implication of this. I guess that when XTAL will start settle time is short enough and it does not matter
when change form HFINT to HFXO will take place.
On the other hand if XTAL fails to start (I'm not sure what are the odds for this provided that hardware is correct) code will ignore this and will continue to run on HFINT, while before the change, it would stuck in loop and probably watchdog would cause the reboot.
That's why I added @wes3 as the original author who may have some insight.

For me it looks OK

**Comment by wes3** [comment]
@ncasaril and @kasjer I was looking over the changes and I was thinking the same thing regarding settling time. The code to request the hfxo just turns it on and leaves without waiting for it to settle as you say and I was thinking of the implications of that. I think I modeled this code after some code I saw in an SDK from nordic. The idea was to check to see if the clock was started; if so do nothing. If not, clear the started event, turn on the clock, and wait for it to settle. I figured it would be pretty catastrophic if we never see EVENTS_HFCLKSTARTED so that is why it just loops forever and lets the watchdog fire. Yes, that assumes there is a watchdog running :-) There is an errata for the chip that states you could see that event erroneously if the clock takes longer than 400 usecs to settle btw so that would have to be handled differently.

So I guess we just need to decide if we care about using a timer or the radio before the clock settles. I would suspect most settling times for most boards to be in the hundreds of usecs so probably not an issue but who knows? I guess if we were being very cautious and "more correct" (bad term) we would move the code that waits for settling into hfxo_request if the clock is not on. So some thoughts on what we could do if we wanted:
1) Make sure on startup that the HFXO is off. This is to deal with any bootloaders that may have turned it on.
2) In HFXO request, wait to settle if the reference count is zero.
3) Related to 2, if all things are working properly, the HFXO should be off if the reference count is zero. Not sure if we care about that case and want to add an assert there for debug (a debug assert that can get compiled out) but this is not critical.

If both of you feel that there really is no need to wait for it to settle I am fine with the code the way it is.

**Comment by ncasaril** [comment]
@wes3 and @kasjer Thanks for your input. 
I've tried to accomodate your suggestions with moving the wait code into hfxo_request. 
It feels like a good solution and that's now the only place that I can find that enables the HFCLK. 

That said, I think the NRF_CLOCK->HFCLKSTAT isn't used correctly to detect if the HFXO is used. It only checks if the task has been started, not if we're actually running off the HFXO. 

Leave me with it for a bit and I'll see if I can get clarity into how it's meant to work. 

So, not ready for merge just yet.

**Comment by ncasaril** [comment]
@wes3  In the hfxo_request I'm waiting for confirmation that the HFCLK has started. I removed the check on the HFCLKSTAT as it was only checking if the task had been triggered (bit 16) and not if the HFXO was used (bit 0) thus not actually starting the HFXO. 

If you're happy with how it looks now I'll clean up the commits.

**Review by wes3 — COMMENTED** [review]
This all seems good to me but I am not in a position currently to test this on a board.  BTW: I never really checked to see if checking only bit 16 in HFCLKSTAT register is sufficient to see if the HFXO is currently running. I do not know why the code got removed to check that the source was the HFXO (bit 0 being 1) but not a big deal if you tested this and it works.

**Comment by ncasaril** [comment]
Sorry guys for all the back and forth on this one and thanks for the input so far.

I'm now checking both the STATE and the SRC of the HFCLK before starting it in hfxo_request. 

Note that I've not covered the case of using LFCLK_SYNT + HFINT as that combination is restricted by syscfg. 
If you feel this is needed I can add the ifdefs needed for this is hal_system_clock_start.

@andrzej-kaczmarek  I think we need your input to see that we're not breaking anything here.

File: hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c
Comment (danielkucera) [inline hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c:50]: Just out of  curiosity... What is this condition supposed to test?

File: hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c
Comment (danielkucera) [inline hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c:56]: what about using this here?
```
while (!NRF_CLOCK->EVENTS_HFCLKSTARTED) {
}
```

File: hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c
Comment (wes3) [reply hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c:50]: I was speaking to @ncasaril about this. The specification shows two bits in that register. SRC is bit 0 and is 0 for HFINT and 1 for HFXO. The STATE bit is bit 16 and is 0 if HFCLK not running and 1 if running. So it seemed best to check both bits being 1. It might be possible to just check SRC and if 1 all is good. Looking at one of the versions of the SDK from nordic it does appear the check both bits.

Comment (danielkucera) [reply hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c:50]: Okay, then this is not very obvious... I would suggest comparing to `CLOCK_HFCLKSTAT_SRC_Xtal` and `CLOCK_HFCLKSTAT_STATE_Running` defines, something like:
```
        if ((NRF_CLOCK->HFCLKSTAT &
             (CLOCK_HFCLKSTAT_SRC_Msk | CLOCK_HFCLKSTAT_STATE_Msk)) !=
            (CLOCK_HFCLKSTAT_SRC_Xtal << CLOCK_HFCLKSTAT_SRC_Pos | 
             CLOCK_HFCLKSTAT_STATE_Running << CLOCK_HFCLKSTAT_STATE_Pos)) {
```

Comment (ncasaril) [reply hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c:50]: I agree that this is clearer. The reason we used the other if-statment was that this was the original statement used and I wanted to keep the changes to a minimum. I'll change so it's clearer.

File: hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c
Comment (ncasaril) [reply hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c:56]: Again, same reason - minimum changes left this one in there. Very happy to simplify this.

**Review by kasjer — APPROVED** [review]
LGTM, if no one has anything to add soon I will merge it very soon.

File: hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c
Comment (danielkucera) [inline hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c:54]: this logic is reversed - `NRF_CLOCK->EVENTS_HFCLKSTARTED` has to be non-zero to escape the loop

Comment (kasjer) [reply hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c:54]: thanks @danielkucera

**Review by kasjer — CHANGES_REQUESTED** [review]

File: hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c
Comment (ncasaril) [reply hw/mcu/nordic/nrf52xxx/src/nrf52_clock.c]: Dang, too quick. Thanks guys.

**Review by kasjer — APPROVED** [review]
LGTM

**Pull request merged**
